File: app/src/main/java/com/annapol04/munchkin/engine/PlayClient.java
package com.annapol04.munchkin.engine;

import android.app.Activity;
import android.content.Intent;
import android.util.Log;

import java.util.Arrays;

public abstract class PlayClient {
    public enum MatchState {
        LOGGED_OUT,
        LOGGED_IN,
        STARTED,
        ABORTED,
        DISCONNECTED,
    }

    public interface OnMatchStateChangedListener {
        void onMatchStateChanged(MatchState state);
    }

    public interface OnMessageReceivedListener {
        void onMessageReceived(byte[] data);
    }

    protected OnMatchStateChangedListener matchStateChangedListener;
    protected OnMessageReceivedListener messageReceivedListener;

    public void setMatchStateChangedListener(OnMatchStateChangedListener matchStateChangedListener) {
        this.matchStateChangedListener = matchStateChangedListener;
    }

    protected void changeMatchState(MatchState state) {
        if (matchStateChangedListener != null)
            matchStateChangedListener.onMatchStateChanged(state);
    }

    public void setMessageReceivedListener(OnMessageReceivedListener messageReceivedListener) {
        this.messageReceivedListener = messageReceivedListener;
    }

    protected void messageReceived(byte[] data) {
        if (messageReceivedListener != null)
            messageReceivedListener.onMessageReceived(data);
    }

    public abstract void setActivity(Activity activity);

    public abstract boolean isLoggedIn();

    public abstract void login();

    public abstract void processActivityResults(int requestCode, int resultCode, Intent data);

    public abstract void startQuickGame();

    public abstract void sendToAll(byte[] message);

    public abstract int getAmountOfPlayers();
}

File: app/src/main/java/com/annapol04/munchkin/engine/PlayClient.kt
package com.annapol04.munchkin.engine

import android.app.Activity
import android.content.Intent
import android.util.Log

abstract class PlayClient {
    val TAG = PlayClient::class.java.simpleName

    var matchStateChangedListener: OnMatchStateChangedListener? = null
    var messageReceivedListener: OnMessageReceivedListener? = null

    var matchState = ClientState.LOGGED_OUT
        protected set(value) {
            if (field != value) {
                field = value

                Log.i(TAG, "MATCH_STATE=" + value.toString())

                matchStateChangedListener?.onMatchStateChanged(value)
            }
        }

    var errorReason: Int? = null

    abstract val isLoggedIn: Boolean

    abstract val amountOfPlayers: Int

    enum class ClientState {
        LOGGED_OUT,
        LOGGED_IN,
        MATCHMAKING,
        STARTED,
        ABORTED,
        DISCONNECTED
    }

    interface OnMatchStateChangedListener {
        fun onMatchStateChanged(state: ClientState)
    }

    interface OnMessageReceivedListener {
        fun onMessageReceived(data: ByteArray)
    }

    protected fun messageReceived(data: ByteArray) {
        messageReceivedListener?.onMessageReceived(data)
    }

    abstract fun setActivity(activity: Activity?)

    abstract fun login()

    abstract fun processActivityResults(requestCode: Int, resultCode: Int, data: Intent?)

    abstract fun startQuickGame()

    abstract fun leaveGame()

    abstract fun sendToAll(message: ByteArray)
}
